libvast/src/system/index.cpp

```diff
@@ -1236,15 +1235,22 @@ index(index_actor::stateful_pointer<index_state> self,
                   self->state.partition_synopsis_path(new_partition_id))
         .then(
           [self, rp, old_partition_ids, new_partition_id,
-           keep](partition_synopsis_ptr& synopsis) mutable {
-            auto result = partition_synopsis_pair{new_partition_id, synopsis};
-            // TODO: We eventually want to allow transforms that delete
-            // whole events, at that point we also need to update the index
-            // statistics here.
+           keep](augmented_partition_synopsis& aps) mutable {
+            auto result = partition_info{
+              .uuid = aps.uuid,
+              .events = aps.synopsis->events,
+              .max_import_time = aps.synopsis->max_import_time,
+              .stats = std::move(aps.stats),
+            };
+            // Update the index statistics. We only need to add the events of
+            // the new partition here, the subtraction of the old events is
+            // done in `erase`.
+            for (const auto& [name, stats] : result.stats.layouts)
+              self->state.stats.layouts[name].count += stats.count;
```

Comment:
I'd argue this becomes a problem now. If the `erase` message gets delayed because of a large backlog intermediate requests will produce incorrect results. Shouldn't the erase logic be factored into a function at this point?

Reply:
Hm, it's an interesting question what the "correct" number would be, because the events will still exist on disk until the `erase` went through, even though they're not reachable through the meta index anymore.

The erase handler goes through a number of continuations itself, so it would be non-trivial to refactor it into a function. I think the best alternative would probably be to return two sets of `index_statistics` from the partition transformer, one for the incoming and one from the outgoing events, and then to make a second erase handler that does not update statistics or remove from the meta index.

Reply:
I'm afraid we're thinking about this at the wrong level. The ultimate source of truth is the meta index because that controls which partitions are reachable. I think in the long run we should move the statistics for the passive database to the meta index and sum up it up with the active statistics in the node.

Reply:
This should result in a follow-up story. Let's not make that change in this PR.

Reply:
I agree this should happen in a different PR.

Reply:
I created https://app.shortcut.com/tenzir/story/31848/atomically-adjust-index-statistics-for-partition-transforms